rlp/list.libevm_test.go

```diff
@@ -54,3 +58,195 @@ func TestDecodeList(t *testing.T) {
 		assert.Equalf(t, vals[i], *gotPtr, "DecodeList()[%d]", i)
 	}
 }
+
+func TestStructFieldHelpers(t *testing.T) {
+	type foo struct {
+		A uint64
+		B uint64
+		C *uint64
+		D *uint64   `rlp:"optional"`
+		E []uint64  `rlp:"optional"`
+		F *[]uint64 `rlp:"optional"`
+	}
+
+	const (
+		a uint64 = iota
+		b
+		cVal
+		dVal
+	)
+	c := common.PointerTo(cVal)
+	d := common.PointerTo(dVal)
+	e := []uint64{40, 41}
+	f := &[]uint64{50, 51}
+
+	tests := []foo{
+		{a, b, c, d, e, f},       // 000 (which of d/e/f are nil)
+		{a, b, c, d, e, nil},     // 001
+		{a, b, c, d, nil, f},     // 010
+		{a, b, c, d, nil, nil},   // 011
+		{a, b, c, nil, e, f},     // 100
+		{a, b, c, nil, e, nil},   // 101
+		{a, b, c, nil, nil, f},   // 110
+		{a, b, c, nil, nil, nil}, // 111
+		// Empty and nil slices are treated differently when optional
+		{a, b, c, nil, []uint64{}, nil},
+		{a, b, c, nil, nil, &[]uint64{}},
+	}
+
+	for _, obj := range tests {
+		obj := obj
+		t.Run("", func(t *testing.T) {
+			t.Logf("\n%s", pretty.Sprint(obj))
+
+			wantRLP, err := EncodeToBytes(obj)
+			require.NoErrorf(t, err, "EncodeToBytes([actual struct])")
+
+			t.Run("EncodeStructFields", func(t *testing.T) {
+				var got bytes.Buffer
+				err = EncodeStructFields(
+					&got,
+					[]any{obj.A, obj.B, obj.C},
+					[]any{obj.D, obj.E, obj.F},
+				)
+				require.NoErrorf(t, err, "EncodeStructFields(..., [required], [optional])")
+
+				assert.Equal(t, wantRLP, got.Bytes(), "EncodeToBytes() vs EncodeStructFields()")
+			})
+
+			t.Run("DecodeStructFields", func(t *testing.T) {
+				s := NewStream(bytes.NewReader(wantRLP), 0)
+				var got foo
+				err := s.DecodeStructFields(
+					[]any{&got.A, &got.B, &got.C},
+					[]any{&got.D, &got.E, &got.F},
+				)
+				require.NoError(t, err, "Stream.DecodeStructFields(...)")
+
+				var want foo
+				err = DecodeBytes(wantRLP, &want)
+				require.NoError(t, err, "DecodeBytes(...)")
+
+				assert.Equal(t, want, got, "DecodeBytes(..., [original struct]) vs Stream.DecodeStructFields(...)")
+			})
+		})
+	}
+}
+
+//nolint:testableexamples // Demonstrating code equivalence, not outputs.
```

Comment:
nit does this `//nolint` actually do anything? I thought it applied to golangci-lint, but testableexamples isn't a linter in golangci-lint 🤔

Reply:
They just don't seem to have it documented. Linting without this:

```
rlp/list.libevm_test.go:135:41: missing output for example, go test can't validate it (testableexamples)
func ExampleStream_DecodeStructFields() {
```

It would also be flagged if unnecessary. Running on:

```go
//nolint:testableexamples
func Example() {
	fmt.Println("foo")
	// Output: foo
}
```

results in:

```
rlp/list.libevm_test.go:136:1: directive `//nolint:testableexamples` is unused for linter "testableexamples" (nolintlint)
//nolint:testableexamples
```